Create rule S6994 Use static_assert for compile-time constants CPP-5134 (#4025)
SonarSource/rspec: rules/S6994/cfamily/rule.adoc

== Why is this an issue?

> Introduced in C++11, `static_assert` will check a precondition at compile-time and emit a diagnostic error message.

Reviewer: I'm surprised our tests did not detect these issues. (`static_assert` needs to be updated in the rest of the text too.)

```suggestion
Introduced in {cpp}11, ``++static_assert++`` will check a precondition at compile-time and emit a diagnostic error message.
```

I'd also suggest using the present tense for facts:

```suggestion
Introduced in {cpp}11, ``++static_assert++`` checks a precondition at compile-time and emits a diagnostic error message.
```

Reviewer 2: `static_assert` is OK. The two things we will detect are two consecutive '_' or '_' at the beginning or end of a word. A single '_' inside a word will never turn into a tag.

Reviewer: Right, I forgot about this quirk of the language...

For any condition that could be checked at compile-time, `static_assert` should be preferred to the C macro `assert`:

> * Because it is checked at compile-time, the maintainer changing the code will be warned immediately in case of condition breakage, as opposed to `assert` which will trigger only when the specific line is run, potentially by another person.

Reviewer: ```suggestion
* Because it is checked at compile-time, the maintainer changing the code is immediately warned when the condition breaks, as opposed to `assert` which will trigger only when the specific line is run, potentially by another person.
```

* Because it is checked at compile-time, it is impossible to accidentally create a side-effect.

> * It cannot create a runtime performance cost.

Reviewer: ```suggestion
* It cannot have a runtime performance cost.
```

* Contrary to `assert`, it is not a macro and thus doesn't have the same surprising behaviors like bad handling of commas or confusing error messages when it doesn't compile.

* It is designed to emit a customizable error message which will be useful to diagnose the error.

[source,cpp,diff-id=1,diff-type=noncompliant]
----
template <class T>
void testFunction(T t) {

>   assert (sizeof(long) <= sizeof(long long) && "long long is smaller than long!"); // Noncompliant
>   assert (std::is_integral_v<T> && "This template only works for integral types"); // Noncompliant

Reviewer: For consistency
```suggestion
  assert(sizeof(long) <= sizeof(long long) && "long long is smaller than long!"); // Noncompliant
  assert(std::is_integral_v<T> && "This template only works for integral types"); // Noncompliant
```

  assert(t > 0); // Compliant: only known at runtime
  // ...
}
----

[source,cpp,diff-id=1,diff-type=compliant]
----
template <class T>
void testFunction(T t) {

>   static_assert (sizeof(long) <= sizeof(long long), "long long is smaller than long!");
>   static_assert (std::is_integral_v<T>, "This template only works for integral types");

Reviewer: ```suggestion
  static_assert(sizeof(long) <= sizeof(long long), "long long is smaller than long!");
  static_assert(std::is_integral_v<T>, "This template only works for integral types");
```

  assert(t > 0); // Compliant: only known at runtime
  // ...
}
----

This rule doesn't trigger for assertions that are false on purpose to stop the execution flow:

[source,cpp]
----
[[noreturn]] void stop() {
  assert(false); // Compliant: Interrupting the runtime execution
  assert(!"We shouldn't reach this"); // Compliant: Interrupting the runtime execution
}
----

== Resources

> 

Reviewer: I would also add a link to `static_assert` and `assert` on C++ reference.
